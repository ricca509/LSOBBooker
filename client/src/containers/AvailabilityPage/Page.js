import React from 'react';
import { observer, inject } from 'mobx-react';
import { first, last } from 'lodash';

import AvailabilityList from '../AvailabilityPage/AvailabilityList';
import LocationHeader from '../AvailabilityPage/LocationHeader';
import Loading from '../../components/Loading';
import { setLocation, setService, resetLocationAndService } from '../../actions';

const onSelectedLocationChange = (ev) => {
  setLocation(ev.target.value);
}

const onSelectedServiceChange = (ev) => {
  setService(ev.target.value);
}

const onResetClick = () => {
  resetLocationAndService();
}

const getDates = (hasAvailability, availability) => {
  return hasAvailability ?
    {
      start: first(availability)['date'],
      end: last(availability)['date']
    } :
    null;
}

const renderAvailability = ({
  fetching,
  fetched,
  fetchError,
  availability
}) => {
  if (fetching) {
    return <Loading />;
  }

  if (fetched) {
    return <AvailabilityList availability={availability} />;
  }

  if (fetchError) {
    return <div>Fetch error</div>;
  }
}

const renderHeader = ({
  fetchError,
  selectedLocationId,
  selectedEventId,
  hasAvailability,
  availability
}) => {
  if (fetchError) return null;

  return (
    <div className="row">
      <LocationHeader
        onSelectedLocationChange={onSelectedLocationChange}
        onSelectedServiceChange={onSelectedServiceChange}
        onResetClick={onResetClick}
        locationId={selectedLocationId}
        serviceId={selectedEventId}
        dates={getDates(hasAvailability, availability)}
      />
  </div>
  );
}

export const Page = ({ store }) => (
  <div className="container">
    <div className="row">
      <h3>London School of Barbering availability</h3>
    </div>
      { renderHeader(store) }
    <div className="row">
      { renderAvailability(store) }
    </div>
  </div>
);

Page.displayName = 'Page';

export default inject('store')(observer(Page));
