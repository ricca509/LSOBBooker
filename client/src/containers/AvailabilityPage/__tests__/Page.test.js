import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { Page } from '../Page';

describe('Page', () => {
  describe('when there is availability', () => {
    it('renders the data', () => {
      const wrapper = shallow(<Page
        availability={[{
          date: 'date',
          times: [1, 2]
        }]}
        hasAvailability={true}
        locationId={4}
        eventId={68}
      />);

      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe('when there is no availability', () => {
    it('renders the loading spinner', () => {
      const wrapper = shallow(<Page
        availability={[]}
        hasAvailability={false}
        locationId={4}
        eventId={68}
      />);

      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });
});
