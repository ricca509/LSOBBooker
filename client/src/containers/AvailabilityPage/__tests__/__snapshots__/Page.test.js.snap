exports[`Page when there is availability renders the data 1`] = `
<div
  className="container">
  <div
    className="row">
    <h3>
      London School of Barbering availability
    </h3>
  </div>
  <div
    className="row">
    <LocationHeader
      dates={
        Object {
          "end": "date",
          "start": "date",
        }
      }
      location="St Pauls/Farringdon - 6 West Smithfield EC1A 9JX"
      service="1 Fade Cut - 1/4cm on sides" />
  </div>
  <div
    className="row">
    <AvailabilityList
      availability={
        Array [
          Object {
            "date": "date",
            "times": Array [
              1,
              2,
            ],
          },
        ]
      } />
  </div>
</div>
`;

exports[`Page when there is no availability renders the loading spinner 1`] = `
<div
  className="container">
  <div
    className="row">
    <h3>
      London School of Barbering availability
    </h3>
  </div>
  <div
    className="row">
    <LocationHeader
      dates={null}
      location="St Pauls/Farringdon - 6 West Smithfield EC1A 9JX"
      service="1 Fade Cut - 1/4cm on sides" />
  </div>
  <div
    className="row">
    <Loading />
  </div>
</div>
`;
